internal/controller/collection/reconciler.go

```diff
@@ -0,0 +1,277 @@
+	if err != nil {
+		collection.Status.SetConditions(xpv1.ReconcileError(err))
+		if err := r.client.Status().Update(ctx, collection); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+		}
+		return ctrl.Result{}, errors.Wrap(err, errNewKubernetesClient)
+	}
+
+	// Fetch objects based on the set GVK and selector.
+	objects := &unstructured.UnstructuredList{}
+	objects.SetAPIVersion(collection.Spec.APIVersion)
+	objects.SetKind(collection.Spec.Kind)
+	selector, err := metav1.LabelSelectorAsSelector(&collection.Spec.Selector)
+
+	if err != nil {
+		collection.Status.SetConditions(xpv1.ReconcileError(err))
+		if err := r.client.Status().Update(ctx, collection); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+		}
+		return ctrl.Result{}, errors.Wrap(err, "error creating selector")
+	}
+
+	if err := clusterClient.List(ctx, objects, &client.ListOptions{LabelSelector: selector, Namespace: collection.Spec.Namespace}); err != nil {
+		collection.Status.SetConditions(xpv1.ReconcileError(err))
+		if err := r.client.Status().Update(ctx, collection); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+		}
+		return ctrl.Result{}, errors.Wrap(err, "error listing objects")
+	}
+
+	// Create observed-only Objects for all found items
+	refs := sets.New[v1alpha1.ObservedObjectReference]()
+	for _, o := range objects.Items {
+		log.Debug("creating observed object for the matched item", "gvk", o.GroupVersionKind(), "name", o.GetName())
+		name, err := r.observedObjectName(collection, &o) //#nosec G601
+		if err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("error generating name for observed object, matched object: %v", o))
+		}
+
+		// Create patch
+		po, err := observedObjectPatch(name, o, collection)
+		if err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("error generating patch for matched object %v", o))
+		}
+
+		if err := r.client.Patch(ctx, po, client.Apply, fieldOwner, client.ForceOwnership); err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, "cannot create observed object")
+		}
+
+		log.Debug("created observed object", "name", po.GetName())
+		refs.Insert(v1alpha1.ObservedObjectReference{Name: name})
+	}
+
+	// Remove objects that either do not exist anymore or are no match
+	oldRefs := sets.New[v1alpha1.ObservedObjectReference](collection.Status.Objects...)
```

Comment:
Currently, we have two distinct sources (`status.objects` and objects owned by collection on etcd), and they could diverge for some reason.

I believe it would be more reliable if we lean on whatever on the API instead of using references in the status (also relying on status has its downsides). I mean, we could add labels to the created observe-only object resources (e.g. kubernetes.crossplane.io/owned-by-collection: jobs) and list them with labels to see what a collection owns and clean them if they no longer needed.

Reply:
If we introduce such label, then we implicitly limit collection names to 63 chars. Let me know if this is an important trade to be made. Also, `ObservedObjectCollection` instances are gonna be created mostly from within a composition, where their name is built from composite name and a prefix.

Reply:
> I believe it would be more reliable if we lean on whatever on the API instead of using references in the status 

Having refs in the status might be helpful for some, similar to how we have refs in the composite object.

Reply:
Yes, agree having refs in the status makes sense for visibility. But we should also add labels so that functions can avoid having to first get the collection and then requesting objects if they already know the collection, they should be able to request resources by label directly instead.

Reply:
>  if they already know the collection, they should be able to request resources by label directly instead.

Users can do that already by adding an arbitrary number of labels through `.spec.template.metadata.labels` on ObserverdObjectCollection.

Reply:
Fair enough, then I'm fine either way.

Reply:
> If we introduce such label, then we implicitly limit collection names to 63 chars. 

>  Also, ObservedObjectCollection instances are gonna be created mostly from within a composition, where their name is built from composite name and a prefix.

This is a good point. However, the same limitation holds for composites, doesn't it? They put `crossplane.io/composite: <name-of-the-composite>` to all resources they compose and they could be created from within other compositions as well. 

I believe the reason why we maintain references there is that the composed resources could be of any apiVersion/kind and it is not possible to list all composed objects with a label filter. The situation is different here, we know all owned resources are `Objects` and we can list/filter them with labels. 

I am totally fine with keeping them in status for visibility.

Reply:
I have introduced `kubernetes.crossplane.io/owned-by-collection` label that contains the name of the collection, if the name length is less than 63 chars. If greater, md5 hash of the name is used instead.

The label and its value is published under `status.membershipLabel` so that clients can know what label to use to obtain all items belonging to the given collection.

Reply:
> If greater, md5 hash of the name is used instead.

FWIW there's a potential collision here:

* One collection has a long enough name that its objects refer to it as `some-md5-hash`
* Another collection is actually named `some-md5-hash`

This probably doesn't matter - it would never happen outside of a collection actively trying to cause a collision. 

Is it so bad to just limit collection names to 63 characters? I prefer predictable and human-readable labels to labels that are sometimes human readable, and other times an md5 hash. I think this would also allow us to avoid having to write a label to the status of the collection just to know what label to query for (e.g. hash or name).